### PR TITLE
Clear stream before emitting error

### PR DESCRIPTION
**Pull request by zbarbuto**
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)

Bug fix

### :arrow_heading_down: What is the current behavior?

When the location stream times out, if your handler for this error is to immediately re-try listening to the stream, you receive a dead stream.

### :new: What is the new behavior (if this is a feature change)?

Clear out the stale stream before throwing the error so re-listening will get a fresh stream.

### :boom: Does this PR introduce a breaking change?

No

### :bug: Recommendations for testing

Something along these lines (but force it to timeout)

```dart
void _setupLocationStream() {
  final stream = Geolocator.getPositionStream(
    locationSettings: LocationSettings(timeLimit: Duration(seconds: 3))
  );
  locationSubscription = stream.listen((location) {
    print(location);
  });
   locationSubscription!.onError((error) {
     // This attempt to re-setup the stream will get the already timed out stream, meaning
     // you do not get any more updates.
     // If you await a Future.delayed(Duration(milliseconds: 100)); here, you do get updates
     _setupLocationStream();
   });
}
```

### :memo: Links to relevant issues/docs

Specifically [this check](https://github.com/NextFaze/flutter-geolocator/blob/master/geolocator_platform_interface/lib/src/implementations/method_channel_geolocator.dart#L156-L158) is not null because the call to check if the existing stream is null fails (verified with breakpoints)

### :thinking: Checklist before submitting

- [x] I made sure all projects build.
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy](https://dart.dev/tools/pub/versioning).
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I followed the style guide lines ([code style guide](https://github.com/Baseflow/flutter-geolocator/blob/master/CONTRIBUTING.md)).
- [ ] I updated the relevant documentation.
- [x] I rebased onto current `master`.

### CODE REVIEW

File: geolocator_platform_interface/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,6 @@
+## 4.0.4
+
+- Fix listening location to stream immediately on error
```
Comment (mvanbeusekom) [inline geolocator_platform_interface/CHANGELOG.md:3]: We try to write change log entries starting with a third person singular verb. How about something like:

```suggestion
- Fixes a bug where listening to the position stream immediately after an error, results in listening to a dead stream. 
```

**Review by mvanbeusekom — CHANGES_REQUESTED** [review]
Hi @zbarbuto,

Thank you for this contribution, good catch. 

Could you also provide a unit test that proves the previous situation is broken and this fix resolves the issue? That will help us to prevent the same bug in the future.

File: geolocator_platform_interface/CHANGELOG.md (hunk 1 above)
Comment (zbarbuto) [reply geolocator_platform_interface/CHANGELOG.md]: Done, thanks

**Comment by zbarbuto** [comment]
Have added a test and squashed commits as cleanup.

Sorry the test is a bit clumsy, was a bit complex to reproduce the issue without the fix in testing environment.

**Review by mvanbeusekom — APPROVED** [review]
LGTM, thanks again for this contribution

**Pull request merged**